src/consensus/encode.rs

```diff
@@ -522,21 +619,52 @@ impl<T: Encodable> Encodable for Box<[T]> {
 
 impl<T: Decodable> Decodable for Box<[T]> {
     #[inline]
-    fn consensus_decode<R: io::Read + ?Sized>(r: &mut R) -> Result<Self, Error> {
-        let len = VarInt::consensus_decode(r)?.0;
-        let len = len as usize;
+    fn consensus_decode<R: io::Read + ?Sized + Seek>(
+        r: &mut R,
+        bytes_upper_limit: usize,
+    ) -> Result<Self, EncodeError> {
+        let len = usize::try_from(*VarInt::consensus_decode(r, mem::size_of::<VarInt>() * 2)?)
+            .map_err(|e| {
+                self::EncodeError::ParseFailed(format!("VarInt overflows usize ({})", e))
+            })?;
+
+        // Prevent allocations larger than the maximum allowed size
+        let layout_size = mem::size_of::<T>().saturating_mul(len);
+        if layout_size > max_allocatable_mem_bytes_size(bytes_upper_limit) {
+            return Err(self::EncodeError::ParseFailed(format!(
+                "length ({} x {} = {}) exceeds maximum allocatable bytes ({}) by {} bytes",
+                mem::size_of::<T>(),
+                len,
+                layout_size,
+                max_allocatable_mem_bytes_size(bytes_upper_limit),
+                layout_size.saturating_sub(max_allocatable_mem_bytes_size(bytes_upper_limit)),
+            )));
+        }
+
         let mut ret = Vec::with_capacity(len);
         for _ in 0..len {
-            ret.push(Decodable::consensus_decode(r)?);
+            ret.push(Decodable::consensus_decode(r, bytes_upper_limit)?);
         }
         Ok(ret.into_boxed_slice())
     }
 }
 
+// Maximum allocatable memory bytes size - this number does not need to be exact, it only needs to be ample, as its
+// primary purpose is to prevent memory overflows during consensus decoding for vectors or arrays.
+pub(crate) fn max_allocatable_mem_bytes_size(bytes_upper_limit: usize) -> usize {
+    const MAX_MEM_ALLOC_SIZE: usize = 32 * 1024 * 1024; // 32 MiB
+    min(MAX_MEM_ALLOC_SIZE, max(bytes_upper_limit, 128))
+}
+
```

Comment:
Continuing the other comment if `MAX_MEM_ALLOC_SIZE` is the max then why not just always use the max?